synapse_spamcheck_badlist/BadListFilter.py

```diff
@@ -0,0 +1,212 @@
+import linkify_it
+from linkify_it import LinkifyIt
+from linkify_it.tlds import TLDS
+import hashlib
+import logging
+import re
+from urllib.parse import urlparse, urlunparse, quote as urlquote
+
+logger = logging.getLogger("synapse.spamfilter.badlist." + __name__)
+
+class BadListFilter(object):
+    """
+    A simple spam checker module for Synapse, designed to block upload of identified child sexual abuse
+    imagery and links to identified child sexual abuse websites.
+
+    This filter requires:
+    - a database of links of identified child sexual abuse websites (as published by e.g. the IWF);
+    - a database of MD5s of identified child sexual abuse imagery (as published by e.g. the IWF).
+
+    The filter:
+    - rejects any message containing a link that matches the database;
+    - rejects any upload containing a file that matches the database.
+    """
+    def __init__(self, config, api):
+        # The plug-in API.
+        self._api = api
+
+        # The table containing links. Configured in homeserver.yaml, spam_checker.config.links_table.
+        self._links_table = config["links_table"]
+        logger.info("Using links table %s" % self._links_table)
+
+        # The table containing md5 hashes. Configured in homeserver.yaml, spam_checker.config.links_table.
+        self._md5_table = config["md5_table"]
+        logger.info("Using md5 table %s" % self._md5_table)
+
+        # The base url for this server. Configured in homeserver.yaml, spam_checker.config.base_url.
+        self._base_url = config["base_url"]
+        logger.info("Using base url %s" % self._base_url)
+
+        # Regexp for extracting info from mxc links.
+        self._mxc_re = re.compile("mxc://(?P<server_name>.*)/(?P<media_id>.*)")
```

Comment:
It seems that Synapse has a slightly different regex: https://github.com/matrix-org/synapse/blob/895e04319ba457a855207b8bb804f7360a258464/synapse/storage/databases/main/room.py#L663

Not sure if that's useful or not.